server/commands/write.go

```diff
@@ -87,6 +90,58 @@ func (c *WriteCommand) validateTuplesets(ctx context.Context, req *openfgapb.Wri
 	return nil
 }
 
+// validateTypes makes sure that when writing a tuple, the types are compatible.
+// 1. If the tuple is of the form (person:bob, reader, doc:budget), then the type "doc", relation "reader" allows type "person".
+// 2. If the tuple is of the form (group:abc#member, reader, doc:budget), then the type "doc", relation "reader" must allow type "group", relation "member".
+// 3. If the tuple is of the form (*, reader, doc:budget), we allow it only if the type "doc" relation "reader" allows at least one type (with no relation)
+func (c *WriteCommand) validateTypes(ctx context.Context, store string, authorizationModelID string, tk *openfgapb.TupleKey, dbCallsCounter utils.DBCallCounter) error {
+	objectType, _ := tupleUtils.SplitObject(tk.GetObject()) // e.g. "doc"
+
+	dbCallsCounter.AddReadCall()
+	typeDefinition, err := c.datastore.ReadTypeDefinition(ctx, store, authorizationModelID, objectType)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return &tupleUtils.TypeNotFoundError{TypeName: objectType}
+		}
+		return err
+	}
+	relations := typeDefinition.GetMetadata().GetRelations()
+	if relations == nil {
+		// authorization model is old or invalid and does not have type information
+		return nil
```

Comment:
We should be handling the error case where it's invalid differently, right?

For example, if we're using schema_version 1.0 then relations will be nil and we can return nil safely, but if schema_version is 1.1 and relations is nill, then that's a problem. Let's make sure we cover that non-nil error case here.